File: 1962-remove-stones-to-minimize-the-total/1962-remove-stones-to-minimize-the-total.java
import java.util.PriorityQueue;

class Solution {
    public int minStoneSum(int[] piles, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>(piles.length, Collections.reverseOrder());
        int total = 0;
        
        for (int pile : piles) {
            total += pile;
            pq.add(pile);
        }
        
        while (k-- > 0) {
            int temp = pq.poll();
            total -= temp / 2;
            pq.add(temp - temp / 2);
        }
        
        return total;
    }
}

File: 1962-remove-stones-to-minimize-the-total/1962-remove-stones-to-minimize-the-total.py
import heapq
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        heap = []
        total = 0
        
        for pile in piles:
            total += pile
            heapq.heappush(heap, -1 * pile)
        
        while (k > 0):
            temp = -1 * heapq.heappop(heap)
            total -= temp // 2
            heapq.heappush(heap, -1 * (temp - temp // 2))
            k -= 1
        
        return total
            